processor/agentmetricsprocessor/utils_calculate_utilizations.go

```diff
@@ -0,0 +1,263 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package agentmetricsprocessor
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+)
+
+// The following code calculates a new utilization metric from
+// a usage metric across one label (dimension) using the formula:
+//
+// value{l1=v1,...} = value{l1=v1,...} / sum(value{l1=vx,...}) over x=1..N
+
+const (
+	cpuTime         = "system.cpu.time"
+	memoryUsage     = "system.memory.usage"
+	fileSystemUsage = "system.filesystem.usage"
+	swapUsage       = "system.swap.usage"
+)
+
+var metricsToComputeUtilizationFor = map[string]bool{
+	cpuTime:         true,
+	memoryUsage:     true,
+	fileSystemUsage: true,
+	swapUsage:       true,
+}
+
+const stateLabel = "state"
+
+func (mtp *agentMetricsProcessor) appendUtilizationMetrics(rms pdata.ResourceMetricsSlice) error {
+	for i := 0; i < rms.Len(); i++ {
+		ilms := rms.At(i).InstrumentationLibraryMetrics()
+		for j := 0; j < ilms.Len(); j++ {
+			metrics := ilms.At(j).Metrics()
+			for k := 0; k < metrics.Len(); k++ {
+				metric := metrics.At(k)
+
+				// ignore all metrics except the ones we want to compute utilizations for
+				metricName := metric.MetricDescriptor().Name()
+				if _, ok := metricsToComputeUtilizationFor[metricName]; !ok {
+					continue
+				}
+
+				// calculate new utilization metric and append it
+				utilizationMetric, err := mtp.calculateUtilizationMetric(metric)
+				if err != nil {
+					return err
+				}
+
+				metrics.Append(&utilizationMetric)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (mtp *agentMetricsProcessor) calculateUtilizationMetric(metric pdata.Metric) (pdata.Metric, error) {
+	utilizationMetric := pdata.NewMetric()
+	metric.CopyTo(utilizationMetric)
+	utilizationMetric.MetricDescriptor().SetName(metricPostfixRegex.ReplaceAllString(metric.MetricDescriptor().Name(), "utilization"))
+	utilizationMetric.MetricDescriptor().SetType(pdata.MetricTypeDouble)
+
+	// for "cpu.time", we need to convert cumulative values to delta values before
+	// computing utilization of the deltas
+	isCPUTime := metric.MetricDescriptor().Name() == cpuTime
+	if isCPUTime {
+		mtp.convertPrevCPUTimeToDelta(utilizationMetric)
+	}
+
+	if err := calculateUtilizationFromDoubleDataPoints(utilizationMetric); err != nil {
+		return pdata.NewMetric(), err
+	}
+
+	if err := calculateUtilizationFromInt64DataPoints(utilizationMetric); err != nil {
+		return pdata.NewMetric(), err
+	}
+
+	// persist the values of "cpu.time" so we can compute deltas on the next cycle
+	if isCPUTime {
+		mtp.prevCPUTimeValues = doubleDataPointsToMap(metric)
```

Comment:
This mutates processor's state.
Does collector guarantee that a processor will be used in a thread-safe way, i.e. that processor only processes one `ConsumeMetrics` invocation at a time?

Reply:
This isn't guaranteed by the Collector, but in practice will only ever be called once at a time given the way this pipeline is setup (using the host metrics receiver which scrapes metrics once per minute). I'l put this behind a mutex just in case.